src/System.Memory/src/System/Buffers/OwnedMemory.cs

```diff
@@ -0,0 +1,97 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Runtime;
+using System.Runtime.CompilerServices;
+
+namespace System.Buffers
+{
+
+    /// <summary>
+    /// Owner of Memory<typeparamref name="T"/> that provides appropriate lifetime management mechanisms for it.
+    /// </summary>
+    public abstract class OwnedMemory<T> : IDisposable, IRetainable
+    {
+
+        /// <summary>
+        /// The number of items in the Memory<typeparamref name="T"/>.
+        /// </summary>
+        public abstract int Length { get; }
+
+        /// <summary>
+        /// Returns a span wrapping the underlying memory.
+        /// </summary>
+        public abstract Span<T> AsSpan();
+
+        /// <summary>
+        /// Returns a Memory<typeparamref name="T"/> if the underlying memory has not been freed.
+        /// </summary>
+        /// <exception cref="System.ObjectDisposedException">
+        /// Throw when the memory has been disposed.
+        /// </exception>
+        public Memory<T> AsMemory
+        {
+            get 
+            {
+                if (IsDisposed) 
+                {
+                    ThrowHelper.ThrowObjectDisposedException_MemoryDisposed(nameof(OwnedMemory<T>));
+                }
+                return new Memory<T>(this, 0, Length);
+            }
+        }
+
+        /// <summary>
+        /// Returns a handle for the array that has been pinned and hence its address can be taken
+        /// </summary>
+        public abstract MemoryHandle Pin();
+
+        /// <summary>
+        /// Returns an array segment if the underlying memory has not been freed.
```

Comment:
I would remove the "if the memory has not been freed." Otherwise you would have to add it to all the members accessign the buffer.